source/azure_iot.c

```diff
@@ -0,0 +1,438 @@
+        {
+            ulValue2 = ( uint32_t ) ( pcBase64name[ i + 1 ] - '0' ) + 52;
+        }
+        else if( pcBase64name[ i + 1 ] == '+' )
+        {
+            ulValue2 = 62;
+        }
+        else if( pcBase64name[ i + 1 ] == '/' )
+        {
+            ulValue2 = 63;
+        }
+        else
+        {
+            ulValue2 = 0;
+        }
+
+        /* Determine which step we are in.  */
+        if( ulStep == 0 )
+        {
+            /* Use first value and first 2 bits of second value.  */
+            pucName[ j++ ] = ( uint8_t ) ( ( ( ulValue1 & 0x3f ) << 2 ) | ( ( ulValue2 >> 4 ) & 3 ) );
+            i++;
+            ulStep++;
+        }
+        else if( ulStep == 1 )
+        {
+            /* Use last 4 bits of first value and first 4 bits of next value.  */
+            pucName[ j++ ] = ( uint8_t ) ( ( ( ulValue1 & 0xF ) << 4 ) | ( ulValue2 >> 2 ) );
+            i++;
+            ulStep++;
+        }
+        else if( ulStep == 2 )
+        {
+            /* Use first 2 bits and following 6 bits of next value.  */
+            pucName[ j++ ] = ( uint8_t ) ( ( ( ulValue1 & 3 ) << 6 ) | ( ulValue2 & 0x3f ) );
+            i++;
+            i++;
+            ulStep = 0;
+        }
+    }
+
+    /* Put a NULL character in.  */
+    pucName[ j ] = 0;
+    *pulBytesCopied = j;
+
+    return eAzureIoTSuccess;
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Encode Base64 bytes.
+ *
+ * TODO: Remove in favor of embedded SDK implementation.
+ *
+ * */
+static AzureIoTResult_t prvAzureIoTBase64Encode( uint8_t * pucName,
+                                                 uint32_t ulLength,
+                                                 char * pcBase64name,
+                                                 uint32_t base64name_size,
+                                                 uint32_t * pulOutputLength )
+{
+    uint32_t ulPad;
+    uint32_t i, j;
+    uint32_t ulStep;
+
+    /* Adjust the length to represent the base64 name.  */
+    ulLength = ( ( ulLength * 8 ) / 6 );
+
+    /* Default padding to none.  */
+    ulPad = 0;
+
+    /* Determine if an extra conversion is needed.  */
+    if( ( ulLength * 6 ) % 24 )
+    {
+        /* Some padding is needed.  */
+
+        /* Calculate the number of pad characters.  */
+        ulPad = ( ulLength * 6 ) % 24;
+        ulPad = ( 24 - ulPad ) / 6;
+        ulPad = ulPad - 1;
+
+        /* Adjust the ulLength to pickup the character fraction.  */
+        ulLength++;
+    }
+
+    if( base64name_size <= ulLength )
+    {
+        return eAzureIoTOutOfMemory;
+    }
+
+    /* Setup index into the pcBase64name.  */
+    j = 0;
+
+    /* Compute the pcBase64name.  */
+    ulStep = 0;
+    i = 0;
+
+    while( j < ulLength )
+    {
+        /* Determine which step we are in.  */
+        if( ulStep == 0 )
+        {
+            /* Use first 6 bits of name character for index.  */
```

Comment:
What happens if I pass in the string `char* testString = "\xFFThis is a test";`?  From my math, the \0xFF when bit-shifted >>2 will turn into 63.  But the problem is the array we're dereferencing is only 33 so we'll go off the end.

I don't see any guidance about what a legal string is, never mind checking.

Reply:
Just double checking here. The array we index into is `64` long. Can you clarify here?

Reply:
Oops.  Yup, I misread the (33) above and thought it was array len.  This is totally cool as-is.  Sorry about that.